In [1]:
import pandas as pd

# Load the data on referenda
referenda_results_new = pd.read_csv('../data/raw/referenda_results.csv')

# Visualise a summary of the DataFrame
referenda_results_new_info = referenda_results_new.info()
referenda_results_new_head = referenda_results_new.head()
referenda_results_new_info, referenda_results_new_head

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 23886 entries, 0 to 23885
Data columns (total 26 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   id                 23886 non-null  object 
 1   title_de_CH        23886 non-null  object 
 2   title_fr_CH        0 non-null      float64
 3   title_it_CH        0 non-null      float64
 4   title_rm_CH        0 non-null      float64
 5   short_title_de_CH  0 non-null      float64
 6   short_title_fr_CH  0 non-null      float64
 7   short_title_it_CH  0 non-null      float64
 8   short_title_rm_CH  0 non-null      float64
 9   date               23886 non-null  object 
 10  shortcode          23886 non-null  object 
 11  domain             23886 non-null  object 
 12  status             23886 non-null  object 
 13  answer             23886 non-null  object 
 14  type               23886 non-null  object 
 15  ballot_answer      23886 non-null  object 
 16  district           237

(None,
                                                   id  \
 0  bundesgesetz-ueber-eine-sichere-stromversorgun...   
 1  bundesgesetz-ueber-eine-sichere-stromversorgun...   
 2  bundesgesetz-ueber-eine-sichere-stromversorgun...   
 3  bundesgesetz-ueber-eine-sichere-stromversorgun...   
 4  bundesgesetz-ueber-eine-sichere-stromversorgun...   
 
                                          title_de_CH  title_fr_CH  \
 0  Bundesgesetz über eine sichere Stromversorgung...          NaN   
 1  Bundesgesetz über eine sichere Stromversorgung...          NaN   
 2  Bundesgesetz über eine sichere Stromversorgung...          NaN   
 3  Bundesgesetz über eine sichere Stromversorgung...          NaN   
 4  Bundesgesetz über eine sichere Stromversorgung...          NaN   
 
    title_it_CH  title_rm_CH  short_title_de_CH  short_title_fr_CH  \
 0          NaN          NaN                NaN                NaN   
 1          NaN          NaN                NaN                NaN   
 2          NaN  

In [2]:
# Remove columns containing only null values
columns_to_drop = [
    'title_fr_CH', 'title_it_CH', 'title_rm_CH', 'short_title_de_CH', 
    'short_title_fr_CH', 'short_title_it_CH', 'short_title_rm_CH', 'expats'
]
referenda_results_cleaned = referenda_results_new.drop(columns=columns_to_drop)

# Convert 'date' column to datetime format
referenda_results_cleaned['date'] = pd.to_datetime(referenda_results_cleaned['date'])

# Summary of missing values
missing_values_summary = referenda_results_cleaned.isnull().sum()
referenda_results_cleaned_head = referenda_results_cleaned.head()
missing_values_summary, referenda_results_cleaned_head


(id                   0
 title_de_CH          0
 date                 0
 shortcode            0
 domain               0
 status               0
 answer               0
 type                 0
 ballot_answer        0
 district           126
 name               126
 entity_id            0
 counted              0
 yeas                 0
 nays                 0
 invalid              0
 empty                0
 eligible_voters      0
 dtype: int64,
                                                   id  \
 0  bundesgesetz-ueber-eine-sichere-stromversorgun...   
 1  bundesgesetz-ueber-eine-sichere-stromversorgun...   
 2  bundesgesetz-ueber-eine-sichere-stromversorgun...   
 3  bundesgesetz-ueber-eine-sichere-stromversorgun...   
 4  bundesgesetz-ueber-eine-sichere-stromversorgun...   
 
                                          title_de_CH       date shortcode  \
 0  Bundesgesetz über eine sichere Stromversorgung... 2024-06-09      EV-4   
 1  Bundesgesetz über eine sichere Stromversorgung...

- Columns that were entirely null or not relevant have been dropped.

- The date column has been converted to datetime format.

- There are 126 missing values in the district and name columns.

In [3]:
# Investigate rows with missing 'district' and 'name' values
missing_district_name = referenda_results_cleaned[referenda_results_cleaned['district'].isnull() | referenda_results_cleaned['name'].isnull()]
missing_district_name_info = missing_district_name.info()
missing_district_name_head = missing_district_name.head()
missing_district_name_info, missing_district_name_head


<class 'pandas.core.frame.DataFrame'>
Index: 126 entries, 0 to 14499
Data columns (total 18 columns):
 #   Column           Non-Null Count  Dtype         
---  ------           --------------  -----         
 0   id               126 non-null    object        
 1   title_de_CH      126 non-null    object        
 2   date             126 non-null    datetime64[ns]
 3   shortcode        126 non-null    object        
 4   domain           126 non-null    object        
 5   status           126 non-null    object        
 6   answer           126 non-null    object        
 7   type             126 non-null    object        
 8   ballot_answer    126 non-null    object        
 9   district         0 non-null      object        
 10  name             0 non-null      object        
 11  entity_id        126 non-null    int64         
 12  counted          126 non-null    bool          
 13  yeas             126 non-null    int64         
 14  nays             126 non-null    int64       

(None,
                                                     id  \
 0    bundesgesetz-ueber-eine-sichere-stromversorgun...   
 76   volksinitiative-maximal-10-des-einkommens-fuer...   
 152  volksinitiative-fuer-tiefere-praemien-kostenbr...   
 228  volksinitiative-fuer-freiheit-und-koerperliche...   
 304  volksinitiative-fuer-ein-besseres-leben-im-alt...   
 
                                            title_de_CH       date shortcode  \
 0    Bundesgesetz über eine sichere Stromversorgung... 2024-06-09      EV-4   
 76   Volksinitiative «Maximal 10 % des Einkommens f... 2024-06-09      EV-1   
 152  Volksinitiative «Für tiefere Prämien – Kostenb... 2024-06-09      EV-2   
 228  Volksinitiative «Für Freiheit und körperliche ... 2024-06-09      EV-3   
 304  Volksinitiative «Für ein besseres Leben im Alt... 2024-03-03      EV-1   
 
          domain status    answer      type ballot_answer district name  \
 0    federation  final  accepted  proposal      accepted      NaN  NaN   
 76  

- The district and name columns are entirely missing for these 126 entries.

- These rows have entity_id set to 0, which might indicate a general or non-specific entity (e.g., overall results rather than municipality-specific).

In [4]:
# Drop rows with missing 'district' and 'name' values (entity_id == 0)
referenda_results_cleaned = referenda_results_cleaned[referenda_results_cleaned['entity_id'] != 0]

# Verify the result after dropping
remaining_missing_values_summary = referenda_results_cleaned.isnull().sum()
referenda_results_cleaned_info = referenda_results_cleaned.info()

remaining_missing_values_summary, referenda_results_cleaned_info


<class 'pandas.core.frame.DataFrame'>
Index: 23760 entries, 1 to 23885
Data columns (total 18 columns):
 #   Column           Non-Null Count  Dtype         
---  ------           --------------  -----         
 0   id               23760 non-null  object        
 1   title_de_CH      23760 non-null  object        
 2   date             23760 non-null  datetime64[ns]
 3   shortcode        23760 non-null  object        
 4   domain           23760 non-null  object        
 5   status           23760 non-null  object        
 6   answer           23760 non-null  object        
 7   type             23760 non-null  object        
 8   ballot_answer    23760 non-null  object        
 9   district         23760 non-null  object        
 10  name             23760 non-null  object        
 11  entity_id        23760 non-null  int64         
 12  counted          23760 non-null  bool          
 13  yeas             23760 non-null  int64         
 14  nays             23760 non-null  int64     

(id                 0
 title_de_CH        0
 date               0
 shortcode          0
 domain             0
 status             0
 answer             0
 type               0
 ballot_answer      0
 district           0
 name               0
 entity_id          0
 counted            0
 yeas               0
 nays               0
 invalid            0
 empty              0
 eligible_voters    0
 dtype: int64,
 None)

Cleaned Dataset Summary:

- Total Entries: 23,760
- Columns: 18 
- No Missing Values: All columns are now fully populated.

In [5]:
#Filter data for the date range between 2015 and 2024
referenda_results_filtered = referenda_results_cleaned[
    (referenda_results_cleaned['date'] >= '2018-01-01') &
    (referenda_results_cleaned['date'] <= '2024-12-31')
]

# In 2023 the municipalities of Hemberg, Oberhelfenschwil and Neckertal merged to form the new municipality of Neckertal

# Identify the municipalities to merge
municipalities_to_merge = ['Hemberg', 'Oberhelfenschwil', 'Neckertal']

# Separate the data before 2023 for merging
data_before_2023 = referenda_results_filtered[referenda_results_filtered['date'] < '2023-01-01']
data_from_2023_onwards = referenda_results_filtered[referenda_results_filtered['date'] >= '2023-01-01']

# Aggregate results for the municipalities to be merged into Neckertal before 2023
merged_neckertal_before_2023 = data_before_2023[data_before_2023['name'].isin(municipalities_to_merge)]

# Sum the relevant columns
aggregated_neckertal = merged_neckertal_before_2023.groupby(['id', 'title_de_CH', 'date', 'shortcode', 'domain', 'status', 'answer', 'type', 'ballot_answer'])[['yeas', 'nays', 'invalid', 'empty', 'eligible_voters']].sum().reset_index()

# Add the merged name and entity_id for Neckertal
aggregated_neckertal['name'] = 'Neckertal'
aggregated_neckertal['entity_id'] = merged_neckertal_before_2023['entity_id'].mode()[0]
aggregated_neckertal['district'] = merged_neckertal_before_2023['district'].mode()[0]
aggregated_neckertal['counted'] = merged_neckertal_before_2023['counted'].mode()[0]

# Combine the aggregated data with the rest of the data
data_before_2023_cleaned = data_before_2023[~data_before_2023['name'].isin(municipalities_to_merge)]
referenda_results_final = pd.concat([data_before_2023_cleaned, aggregated_neckertal, data_from_2023_onwards])

# Verify the final result
referenda_results_final_info = referenda_results_final.info()
referenda_results_final_head = referenda_results_final.head()
referenda_results_final_info, referenda_results_final_head


<class 'pandas.core.frame.DataFrame'>
Index: 6450 entries, 1285 to 1283
Data columns (total 18 columns):
 #   Column           Non-Null Count  Dtype         
---  ------           --------------  -----         
 0   id               6450 non-null   object        
 1   title_de_CH      6450 non-null   object        
 2   date             6450 non-null   datetime64[ns]
 3   shortcode        6450 non-null   object        
 4   domain           6450 non-null   object        
 5   status           6450 non-null   object        
 6   answer           6450 non-null   object        
 7   type             6450 non-null   object        
 8   ballot_answer    6450 non-null   object        
 9   district         6450 non-null   object        
 10  name             6450 non-null   object        
 11  entity_id        6450 non-null   int64         
 12  counted          6450 non-null   bool          
 13  yeas             6450 non-null   int64         
 14  nays             6450 non-null   int64    

(None,
                                                      id  \
 1285  anderung-des-bundesgesetzes-ueber-die-alters-u...   
 1286  anderung-des-bundesgesetzes-ueber-die-alters-u...   
 1287  anderung-des-bundesgesetzes-ueber-die-alters-u...   
 1288  anderung-des-bundesgesetzes-ueber-die-alters-u...   
 1289  anderung-des-bundesgesetzes-ueber-die-alters-u...   
 
                                             title_de_CH       date shortcode  \
 1285  Änderung des Bundesgesetzes über die Alters- u... 2022-09-25      EV-3   
 1286  Änderung des Bundesgesetzes über die Alters- u... 2022-09-25      EV-3   
 1287  Änderung des Bundesgesetzes über die Alters- u... 2022-09-25      EV-3   
 1288  Änderung des Bundesgesetzes über die Alters- u... 2022-09-25      EV-3   
 1289  Änderung des Bundesgesetzes über die Alters- u... 2022-09-25      EV-3   
 
           domain status    answer      type ballot_answer    district  \
 1285  federation  final  accepted  proposal      accepted  St. Galle

- The dataset now includes only data from 2018 to 2024.

- The municipalities of Hemberg, Oberhelfenschwil, and Neckertal have been merged into Neckertal for referendums before 2023.

- Total Entries: 6450

- Columns: 18

- All columns are fully populated.

In [6]:
# Validate and correct data types 

# Check data types
data_types_before = referenda_results_final.dtypes

# Define expected data types
expected_dtypes = {
    'id': 'object',
    'title_de_CH': 'object',
    'date': 'datetime64[ns]',
    'shortcode': 'object',
    'domain': 'object',
    'status': 'object',
    'answer': 'object',
    'type': 'object',
    'ballot_answer': 'object',
    'district': 'object',
    'name': 'object',
    'entity_id': 'int64',
    'counted': 'bool',
    'yeas': 'int64',
    'nays': 'int64',
    'invalid': 'int64',
    'empty': 'int64',
    'eligible_voters': 'int64'
}

# Convert columns to expected data types
for column, dtype in expected_dtypes.items():
    referenda_results_final[column] = referenda_results_final[column].astype(dtype)

# Check data types after conversion
data_types_after = referenda_results_final.dtypes
data_types_before, data_types_after


(id                         object
 title_de_CH                object
 date               datetime64[ns]
 shortcode                  object
 domain                     object
 status                     object
 answer                     object
 type                       object
 ballot_answer              object
 district                   object
 name                       object
 entity_id                   int64
 counted                      bool
 yeas                        int64
 nays                        int64
 invalid                     int64
 empty                       int64
 eligible_voters             int64
 dtype: object,
 id                         object
 title_de_CH                object
 date               datetime64[ns]
 shortcode                  object
 domain                     object
 status                     object
 answer                     object
 type                       object
 ballot_answer              object
 district                   object
 nam

The data types of all columns have been validated and corrected as necessary. The columns now have the appropriate data types:

- Object (String) Columns: id, title_de_CH, shortcode, domain, status, answer, type, ballot_answer, district, name

- Datetime Column: date

- Boolean Column: counted

- Numeric Columns: entity_id, yeas, nays, invalid, empty, eligible_voters

In [7]:
# Save the cleaned referenda results data in data/processed
cleaned_referenda_results_path = '../data/processed/cleaned_referenda_results.csv'
referenda_results_final.to_csv(cleaned_referenda_results_path, index=False)

# Confirm the file is saved
import os
os.path.exists(cleaned_referenda_results_path)


True

In [8]:
# Load the file containing data on party strength of the main left-wing party
party_strength_left_path = '../data/raw/party_strength_left.csv'
party_strength_left = pd.read_csv(party_strength_left_path)

In [9]:
# Inspect the dataset
print(party_strength_left.info())
print(party_strength_left.head())


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 92 entries, 0 to 91
Data columns (total 7 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   BFS_NR       92 non-null     int64 
 1   GEBIET_NAME  92 non-null     object
 2   2023         92 non-null     object
 3   2019         92 non-null     object
 4   2015         92 non-null     object
 5   2011         92 non-null     object
 6   2007         92 non-null     object
dtypes: int64(1), object(6)
memory usage: 5.2+ KB
None
   BFS_NR             GEBIET_NAME  2023  2019  2015  2011  2007
0       1           Gesamtschweiz  18.3  16.8  18.8  18.7  19.5
1       5  Grossregion Ostschweiz   ***   ***   ***   ***   ***
2       8           Kanton Glarus  23.4   ***    45  24.6  55.5
3      14     Kanton Schaffhausen  27.4  26.2  28.8  34.6  34.2
4      15  Kanton Appenzell a.Rh.   ***   ***  28.6   ***   ***


In [10]:
# Select only data related to elections of 2015, 2019, and 2023
selected_years = ['2015', '2019', '2023']
party_strength_left = party_strength_left[['BFS_NR', 'GEBIET_NAME'] + selected_years]


In [11]:
# Count the number of unique municipalities
num_municipalities = party_strength_left['GEBIET_NAME'].nunique()
print("Number of unique municipalities:", num_municipalities)

Number of unique municipalities: 92


In [12]:
print(party_strength_left)

    BFS_NR             GEBIET_NAME  2015  2019  2023
0        1           Gesamtschweiz  18.8  16.8  18.3
1        5  Grossregion Ostschweiz   ***   ***   ***
2        8           Kanton Glarus    45   ***  23.4
3       14     Kanton Schaffhausen  28.8  26.2  27.4
4       15  Kanton Appenzell a.Rh.  28.6   ***   ***
..     ...                     ...   ...   ...   ...
87    3427                Wil (SG)  15.2  15.8  15.9
88    3441             Andwil (SG)   8.1   9.1   8.6
89    3442              Gaiserwald    12  10.7   9.5
90    3443             Gossau (SG)    11  11.6  11.7
91    3444               Waldkirch   6.2   4.5   6.9

[92 rows x 5 columns]


In [13]:
# Drop unecessary datapoints (cantons and regions)
regions_to_drop = [
    'Gesamtschweiz', 'Grossregion Ostschweiz', 'Kanton Glarus', 
    'Kanton Schaffhausen', 'Kanton Appenzell a.Rh.', 
    'Kanton Appenzell i.Rh.', 'Kanton St.Gallen', 'Kanton Graubünden', 
    'Kanton Thurgau', 'Wahlkreis St.Gallen', 'Wahlkreis Rorschach', 
    'Wahlkreis Rheintal', 'Wahlkreis Werdenberg', 'Wahlkreis Sarganserland', 
    'Wahlkreis See-Gaster', 'Wahlkreis Toggenburg', 'Wahlkreis Wil'
]

party_strength_left_cleaned = party_strength_left[~party_strength_left['GEBIET_NAME'].isin(regions_to_drop)]

In [14]:
# Inspect data after cleaning
print(party_strength_left_cleaned.info())
print(party_strength_left_cleaned.head())

<class 'pandas.core.frame.DataFrame'>
Index: 75 entries, 17 to 91
Data columns (total 5 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   BFS_NR       75 non-null     int64 
 1   GEBIET_NAME  75 non-null     object
 2   2015         75 non-null     object
 3   2019         75 non-null     object
 4   2023         75 non-null     object
dtypes: int64(1), object(4)
memory usage: 3.5+ KB
None
    BFS_NR      GEBIET_NAME  2015  2019  2023
17    3201     Häggenschwil   7.5   6.5   7.3
18    3202           Muolen   6.2   4.3     5
19    3203  Stadt St.Gallen  28.1  24.4  25.6
20    3204       Wittenbach  16.7  13.2  12.5
21    3211        Berg (SG)  11.2  10.4   9.7


In [15]:
# Convert year columns to numeric types
party_strength_left_cleaned.loc[:, '2015'] = pd.to_numeric(party_strength_left_cleaned['2015'], errors='coerce')
party_strength_left_cleaned.loc[:, '2019'] = pd.to_numeric(party_strength_left_cleaned['2019'], errors='coerce')
party_strength_left_cleaned.loc[:, '2023'] = pd.to_numeric(party_strength_left_cleaned['2023'], errors='coerce')

# Inspect data to verify the changes
print(party_strength_left_cleaned.info())
print(party_strength_left_cleaned.head())

<class 'pandas.core.frame.DataFrame'>
Index: 75 entries, 17 to 91
Data columns (total 5 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   BFS_NR       75 non-null     int64 
 1   GEBIET_NAME  75 non-null     object
 2   2015         75 non-null     object
 3   2019         75 non-null     object
 4   2023         75 non-null     object
dtypes: int64(1), object(4)
memory usage: 3.5+ KB
None
    BFS_NR      GEBIET_NAME  2015  2019  2023
17    3201     Häggenschwil   7.5   6.5   7.3
18    3202           Muolen   6.2   4.3   5.0
19    3203  Stadt St.Gallen  28.1  24.4  25.6
20    3204       Wittenbach  16.7  13.2  12.5
21    3211        Berg (SG)  11.2  10.4   9.7


The data cleaning process for the left-wing party strength dataset has been successful. The dataset now has the following structure:

- 75 Entries: Corresponding to the municipalities of interest

- Columns: BFS_NR, GEBIET_NAME, 2015, 2019, 2023

- Data Types: Correctly set to int64 for BFS_NR, object for GEBIET_NAME, and float64 for the year columns.

In [16]:
# Save the cleaned and corrected data in data/processed
party_strength_left_cleaned.to_csv('../data/processed/party_strength_left_cleaned.csv', index=False)

In [17]:
# Load the file containing data on the main right-wing party
party_strength_right_path = '../data/raw/party_strength_right.csv'
party_strength_right = pd.read_csv(party_strength_right_path)

# Select only data related to elections of 2015, 2019, and 2023
selected_years = ['2015', '2019', '2023']
party_strength_right = party_strength_right[['BFS_NR', 'GEBIET_NAME'] + selected_years]

# Drop specified regions and districts
regions_to_drop = [
    'Gesamtschweiz', 'Grossregion Ostschweiz', 'Kanton Glarus', 
    'Kanton Schaffhausen', 'Kanton Appenzell a.Rh.', 
    'Kanton Appenzell i.Rh.', 'Kanton St.Gallen', 'Kanton Graubünden', 
    'Kanton Thurgau', 'Wahlkreis St.Gallen', 'Wahlkreis Rorschach', 
    'Wahlkreis Rheintal', 'Wahlkreis Werdenberg', 'Wahlkreis Sarganserland', 
    'Wahlkreis See-Gaster', 'Wahlkreis Toggenburg', 'Wahlkreis Wil'
]

party_strength_right_cleaned = party_strength_right[~party_strength_right['GEBIET_NAME'].isin(regions_to_drop)]

# Convert year columns to numeric types
party_strength_right_cleaned.loc[:, '2015'] = pd.to_numeric(party_strength_right_cleaned['2015'], errors='coerce')
party_strength_right_cleaned.loc[:, '2019'] = pd.to_numeric(party_strength_right_cleaned['2019'], errors='coerce')
party_strength_right_cleaned.loc[:, '2023'] = pd.to_numeric(party_strength_right_cleaned['2023'], errors='coerce')

# Inspect the dataset
print(party_strength_right_cleaned.info())
print(party_strength_right_cleaned.head())

<class 'pandas.core.frame.DataFrame'>
Index: 75 entries, 17 to 91
Data columns (total 5 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   BFS_NR       75 non-null     int64 
 1   GEBIET_NAME  75 non-null     object
 2   2015         75 non-null     object
 3   2019         75 non-null     object
 4   2023         75 non-null     object
dtypes: int64(1), object(4)
memory usage: 3.5+ KB
None
    BFS_NR      GEBIET_NAME  2015  2019  2023
17    3201     Häggenschwil  32.4  33.2  35.9
18    3202           Muolen  36.0  34.6  38.2
19    3203  Stadt St.Gallen  21.1  16.5  18.4
20    3204       Wittenbach  34.3  30.1  32.0
21    3211        Berg (SG)  36.6  33.8  32.8


In [18]:
# Save the cleaned data in data/processed
party_strength_right_cleaned.to_csv('../data/processed/party_strength_right_cleaned.csv', index=False)

In [19]:
# Load the dataset on share of the population under 20 years old
share_under_20_path = '../data/raw/share_under_20.csv'
share_under_20 = pd.read_csv(share_under_20_path)

# Inspect the data
print(share_under_20.info())
print(share_under_20.head())


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 92 entries, 0 to 91
Data columns (total 44 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   BFS_NR       92 non-null     int64  
 1   GEBIET_NAME  92 non-null     object 
 2   2022         92 non-null     float64
 3   2021         92 non-null     float64
 4   2020         92 non-null     float64
 5   2019         92 non-null     float64
 6   2018         92 non-null     float64
 7   2017         92 non-null     float64
 8   2016         92 non-null     float64
 9   2015         92 non-null     float64
 10  2014         92 non-null     float64
 11  2013         92 non-null     float64
 12  2012         92 non-null     float64
 13  2011         92 non-null     float64
 14  2010         92 non-null     float64
 15  2009         92 non-null     object 
 16  2008         92 non-null     object 
 17  2007         92 non-null     object 
 18  2006         92 non-null     object 
 19  2005      

In [20]:
# Select only data related to the year 2019
share_under_20_2019 = share_under_20[['BFS_NR', 'GEBIET_NAME', '2019']]

# Drop unnecessary datapoints
regions_to_drop = [
    'Gesamtschweiz', 'Grossregion Ostschweiz', 'Kanton Glarus', 
    'Kanton Schaffhausen', 'Kanton Appenzell a.Rh.', 
    'Kanton Appenzell i.Rh.', 'Kanton St.Gallen', 'Kanton Graubünden', 
    'Kanton Thurgau', 'Wahlkreis St.Gallen', 'Wahlkreis Rorschach', 
    'Wahlkreis Rheintal', 'Wahlkreis Werdenberg', 'Wahlkreis Sarganserland', 
    'Wahlkreis See-Gaster', 'Wahlkreis Toggenburg', 'Wahlkreis Wil'
]

share_under_20_cleaned = share_under_20_2019[~share_under_20_2019['GEBIET_NAME'].isin(regions_to_drop)]

# Convert year column to numeric type
share_under_20_cleaned.loc[:, '2019'] = pd.to_numeric(share_under_20_cleaned['2019'], errors='coerce')


In [21]:
# verify the changes
print(share_under_20_cleaned.info())
print(share_under_20_cleaned.head())

<class 'pandas.core.frame.DataFrame'>
Index: 75 entries, 17 to 91
Data columns (total 3 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   BFS_NR       75 non-null     int64  
 1   GEBIET_NAME  75 non-null     object 
 2   2019         75 non-null     float64
dtypes: float64(1), int64(1), object(1)
memory usage: 2.3+ KB
None
    BFS_NR      GEBIET_NAME  2019
17    3201     Häggenschwil  41.7
18    3202           Muolen  40.1
19    3203  Stadt St.Gallen  28.2
20    3204       Wittenbach  36.3
21    3211        Berg (SG)  38.5


the data cleaning process for the share_under_20 dataset has been successful. The dataset now has the following structure:

- 75 Entries: Corresponding to the municipalities of interest

- Columns: BFS_NR, GEBIET_NAME, 2019

- Data Types: Correctly set to int64 for BFS_NR, object for GEBIET_NAME, and float64 for the year column 2019.


In [22]:
# Save the cleaned data in data/processed
share_under_20_cleaned.to_csv('../data/processed/share_under_20_cleaned.csv', index=False)

In [23]:
# Load the dataset on share of the population over 64 years old
share_over_64_path = '../data/raw/share_over_64.csv'
share_over_64 = pd.read_csv(share_over_64_path)

# Select only data related to the year 2019
share_over_64_2019 = share_over_64[['BFS_NR', 'GEBIET_NAME', '2019']]

# Drop specified regions and districts
regions_to_drop = [
    'Gesamtschweiz', 'Grossregion Ostschweiz', 'Kanton Glarus', 
    'Kanton Schaffhausen', 'Kanton Appenzell a.Rh.', 
    'Kanton Appenzell i.Rh.', 'Kanton St.Gallen', 'Kanton Graubünden', 
    'Kanton Thurgau', 'Wahlkreis St.Gallen', 'Wahlkreis Rorschach', 
    'Wahlkreis Rheintal', 'Wahlkreis Werdenberg', 'Wahlkreis Sarganserland', 
    'Wahlkreis See-Gaster', 'Wahlkreis Toggenburg', 'Wahlkreis Wil'
]

share_over_64_cleaned = share_over_64_2019[~share_over_64_2019['GEBIET_NAME'].isin(regions_to_drop)]

# Convert year column to numeric type
share_over_64_cleaned.loc[:, '2019'] = pd.to_numeric(share_over_64_cleaned['2019'], errors='coerce')

# Display basic information to verify the changes
share_over_64_cleaned_info = share_over_64_cleaned.info()
share_over_64_cleaned_head = share_over_64_cleaned.head()

<class 'pandas.core.frame.DataFrame'>
Index: 75 entries, 17 to 91
Data columns (total 3 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   BFS_NR       75 non-null     int64  
 1   GEBIET_NAME  75 non-null     object 
 2   2019         75 non-null     float64
dtypes: float64(1), int64(1), object(1)
memory usage: 2.3+ KB


In [24]:
#Save the cleaned dataset in data/processed
share_over_64_cleaned.to_csv('../data/processed/share_over_64_cleaned.csv', index=False)
